packages/keyring-controller/src/KeyringController.ts

```diff
@@ -571,34 +610,42 @@ export class KeyringController extends BaseControllerV2<
   async signTypedMessage(
     messageParams: TypedMessageParams,
     version: SignTypedDataVersion,
-  ) {
+  ): Promise<string | Bytes> {
     try {
       const address = normalizeAddress(messageParams.from);
-      if (!address || !isValidHexAddress(address)) {
+      if (!address || !isValidHexAddress(address as Hex)) {
         throw new Error(
           `Missing or invalid address ${JSON.stringify(messageParams.from)}`,
         );
       }
-      const qrKeyring = await this.getOrAddQRKeyring();
-      const qrAccounts = await qrKeyring.getAccounts();
-      if (
-        qrAccounts.findIndex(
-          (qrAddress: string) =>
-            qrAddress.toLowerCase() === address.toLowerCase(),
-        ) !== -1
-      ) {
-        const messageParamsClone = { ...messageParams };
+
+      if ((await this.getAccountKeyringType(address)) === KeyringTypes.qr) {
+        const qrKeyring = await this.getOrAddQRKeyring();
+        const qrAccounts = await qrKeyring.getAccounts();
         if (
-          version !== SignTypedDataVersion.V1 &&
-          typeof messageParamsClone.data === 'string'
+          qrAccounts.findIndex(
+            (qrAddress: string) =>
+              qrAddress.toLowerCase() === address.toLowerCase(),
+          ) !== -1
         ) {
-          messageParamsClone.data = JSON.parse(messageParamsClone.data);
+          return this.#keyring.signTypedMessage(
+            {
+              from: messageParams.from,
+              data:
+                version !== SignTypedDataVersion.V1 &&
+                typeof messageParams.data === 'string'
+                  ? JSON.parse(messageParams.data)
+                  : messageParams.data,
+            },
+            { version },
+          );
         }
-        return this.#keyring.signTypedMessage(messageParamsClone, { version });
       }
 
-      const { password } = this.#keyring;
-      const privateKey = await this.exportAccount(password, address);
+      const privateKey = await this.exportAccount(
+        this.#keyring.password as string,
```

Comment:
What do you think about adding a comment explaining why this is being typecasted?

Reply:
I added a check at the start of this method to early terminate with an error if password is undefined, so now we don't need this cast anymore